python/ray/tests/test_plasma_unlimited.py

```diff
@@ -0,0 +1,160 @@
+        results = []
+        for _ in range(5):
+            results.append(ray.put(np.zeros(400 * MB, dtype=np.uint8)))
+        _check_spilled_mb(address, spilled=1600)
+    finally:
+        ray.shutdown()
+
+
+def test_fallback_when_spilling_impossible_on_get():
+    try:
+        address = _init_ray()
+        x1 = ray.put(np.zeros(400 * MB, dtype=np.uint8))
+        # x1 will be spilled.
+        x2 = ray.put(np.zeros(400 * MB, dtype=np.uint8))
+        _check_spilled_mb(address, spilled=400)
+        # x1 will be restored, x2 will be spilled.
+        x1p = ray.get(x1)
+        _check_spilled_mb(address, spilled=800, restored=400)
+        # x2 will be restored, triggering a fallback allocation.
+        x2p = ray.get(x2)
+        _check_spilled_mb(address, spilled=800, restored=800)
+        del x1p
+        del x2p
+    finally:
+        ray.shutdown()
+
+
+def test_spilling_when_possible_on_get():
+    try:
+        address = _init_ray()
+        x1 = ray.put(np.zeros(400 * MB, dtype=np.uint8))
+        # x1 will be spilled.
+        x2 = ray.put(np.zeros(400 * MB, dtype=np.uint8))
+        _check_spilled_mb(address, spilled=400)
+        # x1 will be restored, x2 will be spilled.
+        ray.get(x1)
+        _check_spilled_mb(address, spilled=800, restored=400)
+        # x2 will be restored, spilling x1.
+        ray.get(x2)
+        _check_spilled_mb(address, spilled=800, restored=800)
+    finally:
+        ray.shutdown()
+
+
+def test_task_unlimited():
+    try:
+        address = _init_ray()
+        x1 = ray.put(np.zeros(400 * MB, dtype=np.uint8))
+        refs = [x1]
+        # x1 is spilled.
+        x2 = ray.put(np.zeros(400 * MB, dtype=np.uint8))
+        x2p = ray.get(x2)
+        sentinel = ray.put(np.zeros(100 * MB, dtype=np.uint8))
+        _check_spilled_mb(address, spilled=400)
+
+        @ray.remote
+        def consume(refs):
+            # round 1: triggers fallback allocation, spilling of the sentinel
+            # round 2: x2 is spilled
+            ray.get(refs[0])
+            # round 1: triggers fallback allocation.
+            return ray.put(np.zeros(400 * MB, dtype=np.uint8))
+
+        # round 1
+        ray.get(consume.remote(refs))
+        _check_spilled_mb(address, spilled=500, restored=400)
+
+        # round 2
+        del x2p
+        ray.get(consume.remote(refs))
+        _check_spilled_mb(address, spilled=900, restored=800)
+        del sentinel
+    finally:
+        ray.shutdown()
+
+
+def test_task_unlimited_multiget_args():
+    try:
+        address = _init_ray()
+        # Too many refs to fit into memory.
+        refs = []
+        for _ in range(10):
+            refs.append(ray.put(np.zeros(200 * MB, dtype=np.uint8)))
+        x2 = ray.put(np.zeros(600 * MB, dtype=np.uint8))
+        x2p = ray.get(x2)
+        _check_spilled_mb(address, spilled=2000)
+
+        @ray.remote
+        def consume(refs):
+            # Should work without thrashing.
+            ray.get(refs)
```

Comment:
@stephanie-wang it seems the "hang on getting multiple refs" behavior we hypothesized does not actually occur (this tests passes), though I'm not sure why.

Reply:
When you check the memory stats output, does it have thrashing ?

Reply:
It's already checked below (restored=2000).

Reply:
Did you try without this line? I think this would consume object store memory and would cause every new object to go to the fallback allocator first:
```
        x2p = ray.get(x2)
```

Reply:
It seems to work fine without that as well (not sure if it makes any difference, since there's not enough memory to fit all the objects simultaneously in both cases).